% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  11pt,
  a4paper,
]{article}
\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math} % this also loads fontspec
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else
  % xetex/luatex font selection
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}
\usepackage{titling}
\usepackage{fancyhdr}
\usepackage{graphicx}
\usepackage{float}
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\usepackage{bookmark}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same}
% Make links footnotes instead of hotlinks:
\DeclareRobustCommand{\href}[2]{#2\footnote{\url{#1}}}
\hypersetup{
  pdftitle={Debugging step: Verify installations for LSF},
  pdfauthor={Antonia Chroni for SJCRH DNB\_BINF\_Core},
  colorlinks=true,
  linkcolor={black},
  filecolor={Maroon},
  citecolor={blue},
  urlcolor={blue},
  pdfcreator={LaTeX via pandoc}}

\title{Debugging step: Verify installations for LSF}
\author{Antonia Chroni for SJCRH DNB\_BINF\_Core}
\date{}

\begin{document}
\maketitle

{
\hypersetup{linkcolor=}
\setcounter{tocdepth}{2}
\tableofcontents
}
\section{Set up}\label{set-up}

\begin{Shaded}
\begin{Highlighting}[]
\FunctionTok{suppressPackageStartupMessages}\NormalTok{(\{}
  \CommentTok{\# 01A\_run\_seurat\_qc.Rmd}
  \FunctionTok{library}\NormalTok{(future)}
  \FunctionTok{library}\NormalTok{(cowplot)}
  \FunctionTok{library}\NormalTok{(devtools)}
  \FunctionTok{library}\NormalTok{(forcats)}
  \FunctionTok{library}\NormalTok{(GGally)}
  \FunctionTok{library}\NormalTok{(stringr)}
  \FunctionTok{library}\NormalTok{(ggpmisc)}
  \FunctionTok{library}\NormalTok{(ggrepel)}
  \FunctionTok{library}\NormalTok{(miQC) }
  \FunctionTok{library}\NormalTok{(flexmix) }\CommentTok{\# to estimate mixtureModel for miQC}
  \FunctionTok{library}\NormalTok{(scater) }
  \FunctionTok{library}\NormalTok{(Seurat) }
  \FunctionTok{library}\NormalTok{(SingleCellExperiment)}
  \FunctionTok{library}\NormalTok{(irlba) }\CommentTok{\# this solves the issue with RunUMAP code chunk}
  \FunctionTok{library}\NormalTok{(scooter)}
  \FunctionTok{library}\NormalTok{(tidyverse)}
  \FunctionTok{library}\NormalTok{(fs) }\CommentTok{\#file system}
  \FunctionTok{library}\NormalTok{(RColorBrewer)}
  
  \CommentTok{\# 02\_run\_SoupX.Rmd}
  \FunctionTok{library}\NormalTok{(future)}
  \FunctionTok{library}\NormalTok{(knitr)}
  \FunctionTok{library}\NormalTok{(SoupX)}
  \FunctionTok{library}\NormalTok{(Seurat)}
  \FunctionTok{library}\NormalTok{(stringr)}
  \FunctionTok{library}\NormalTok{(tidyverse)}
  \FunctionTok{library}\NormalTok{(tinytex)}
  \FunctionTok{library}\NormalTok{(hdf5r)}

  \CommentTok{\# 03\_run\_scDblFinder.Rmd}
  \FunctionTok{library}\NormalTok{(scDblFinder)}
  \FunctionTok{library}\NormalTok{(Seurat)}
  \FunctionTok{library}\NormalTok{(scater)}
  \FunctionTok{library}\NormalTok{(future)}
  \FunctionTok{library}\NormalTok{(tidyverse)}
  \FunctionTok{library}\NormalTok{(grid)}
  \FunctionTok{library}\NormalTok{(knitr)}

  \CommentTok{\# 04\_run\_filter\_object.Rmd}
  \FunctionTok{library}\NormalTok{(devtools)}
  \FunctionTok{library}\NormalTok{(future)}
  \FunctionTok{library}\NormalTok{(Seurat)}
  \FunctionTok{library}\NormalTok{(patchwork)}
  \FunctionTok{library}\NormalTok{(tidyverse)}
  \FunctionTok{library}\NormalTok{(ggthemes)}
  \FunctionTok{library}\NormalTok{(scooter)}
  \FunctionTok{library}\NormalTok{(RColorBrewer)}
  \FunctionTok{library}\NormalTok{(knitr)}

  \CommentTok{\# 05\_run\_summary\_report.Rmd}
  \FunctionTok{library}\NormalTok{(tidyverse)  }
  \FunctionTok{library}\NormalTok{(knitr)}
  \FunctionTok{library}\NormalTok{(patchwork)}

  \CommentTok{\# 01{-}integrative{-}analysis.Rmd}
  \FunctionTok{library}\NormalTok{(future)}
  \FunctionTok{library}\NormalTok{(tidyverse)}
  \FunctionTok{library}\NormalTok{(patchwork)}
  \FunctionTok{library}\NormalTok{(Seurat)}
  \FunctionTok{library}\NormalTok{(SeuratObject)}
  \FunctionTok{library}\NormalTok{(harmony)}
  \FunctionTok{library}\NormalTok{(rliger)}
  \FunctionTok{library}\NormalTok{(RcppPlanc)}
  \FunctionTok{library}\NormalTok{(SeuratWrappers)}
  \FunctionTok{library}\NormalTok{(scooter)}
  \FunctionTok{library}\NormalTok{(reshape2)}
  \FunctionTok{library}\NormalTok{(RColorBrewer)}
  \FunctionTok{library}\NormalTok{(knitr)}
  
  \CommentTok{\# run{-}cell{-}types{-}annotation.R}
  \FunctionTok{library}\NormalTok{(yaml)}
  \FunctionTok{library}\NormalTok{(tidyverse)}
  \FunctionTok{library}\NormalTok{(celldex)}
  
  \CommentTok{\# 01{-}cell{-}types{-}annotation{-}SingleR{-}broad.Rmd}
  \FunctionTok{library}\NormalTok{(tidyverse)}
  \FunctionTok{library}\NormalTok{(Seurat)}
  \FunctionTok{library}\NormalTok{(SingleR)}
  \FunctionTok{library}\NormalTok{(scooter)}
  \FunctionTok{library}\NormalTok{(knitr)}

  
  \FunctionTok{library}\NormalTok{(ShinyCell)}
  \FunctionTok{library}\NormalTok{(shinyhelper)}
  \FunctionTok{library}\NormalTok{(DT)}
  \FunctionTok{library}\NormalTok{(ggdendro)}
  
  \DocumentationTok{\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#}
  \CommentTok{\# sc{-}epigenie}
  \FunctionTok{library}\NormalTok{(AnnotationHub)}
  \FunctionTok{library}\NormalTok{(ensembldb)}
  \FunctionTok{library}\NormalTok{(GenomeInfoDb) }\CommentTok{\#translation between chromosome names}

  \FunctionTok{library}\NormalTok{(GenomicRanges)}
  \FunctionTok{library}\NormalTok{(rtracklayer)}

  \FunctionTok{library}\NormalTok{(Matrix)}
  \FunctionTok{library}\NormalTok{(htmlwidgets)}

  \FunctionTok{library}\NormalTok{(enrichplot)}

  \FunctionTok{library}\NormalTok{(org.Hs.eg.db)}
  \FunctionTok{library}\NormalTok{(org.Mm.eg.db) }\CommentTok{\# Mouse gene annotation (use org.Hs.eg.db for human)}
  
  \CommentTok{\# Motif analysis with Signac}
  \FunctionTok{library}\NormalTok{(JASPAR2020)}
  \FunctionTok{library}\NormalTok{(TFBSTools)}
  \FunctionTok{library}\NormalTok{(BSgenome.Mmusculus.UCSC.mm10)}
  
  \CommentTok{\# Transcription factor footprinting}
  \FunctionTok{library}\NormalTok{(motifmatchr)}
  \FunctionTok{library}\NormalTok{(BSgenome.Hsapiens.UCSC.hg19)}
  
  \FunctionTok{library}\NormalTok{(monocle3)}
  \FunctionTok{library}\NormalTok{(cicero)}
  \FunctionTok{library}\NormalTok{(biovizBase)}
  \FunctionTok{library}\NormalTok{(clusterProfiler)}
  \DocumentationTok{\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#\#}
\NormalTok{  \})}
\end{Highlighting}
\end{Shaded}

\section{Session Info}\label{session-info}

\begin{verbatim}
## R version 4.4.0 (2024-04-24)
## Platform: x86_64-pc-linux-gnu
## Running under: Ubuntu 22.04.4 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 
## LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0
## 
## locale:
## [1] C
## 
## time zone: Etc/UTC
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] grid      stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] clusterProfiler_4.14.6             biovizBase_1.54.0                 
##  [3] cicero_1.3.9                       Gviz_1.46.1                       
##  [5] monocle3_1.4.26                    BSgenome.Hsapiens.UCSC.hg19_1.4.3 
##  [7] motifmatchr_1.28.0                 BSgenome.Mmusculus.UCSC.mm10_1.4.3
##  [9] BSgenome_1.74.0                    BiocIO_1.16.0                     
## [11] Biostrings_2.74.1                  XVector_0.46.0                    
## [13] TFBSTools_1.44.0                   JASPAR2020_0.99.10                
## [15] org.Mm.eg.db_3.20.0                org.Hs.eg.db_3.20.0               
## [17] enrichplot_1.26.6                  htmlwidgets_1.6.4                 
## [19] rtracklayer_1.66.0                 ensembldb_2.30.0                  
## [21] AnnotationFilter_1.30.0            GenomicFeatures_1.58.0            
## [23] AnnotationDbi_1.68.0               AnnotationHub_3.14.0              
## [25] BiocFileCache_2.14.0               dbplyr_2.5.0                      
## [27] ggdendro_0.2.0                     DT_0.34.0                         
## [29] shinyhelper_0.3.2                  ShinyCell_2.1.0                   
## [31] glue_1.8.0                         gridExtra_2.3                     
## [33] R.utils_2.13.0                     R.oo_1.26.0                       
## [35] R.methodsS3_1.8.2                  reticulate_1.37.0                 
## [37] data.table_1.17.8                  SingleR_2.8.0                     
## [39] celldex_1.16.0                     yaml_2.3.10                       
## [41] reshape2_1.4.4                     SeuratWrappers_0.2.0              
## [43] RcppPlanc_2.0.13                   rliger_2.2.1                      
## [45] harmony_1.2.3                      Rcpp_1.0.12                       
## [47] ggthemes_5.1.0                     patchwork_1.3.2                   
## [49] scDblFinder_1.20.2                 hdf5r_1.3.10                      
## [51] tinytex_0.57                       SoupX_1.6.2                       
## [53] knitr_1.50                         RColorBrewer_1.1-3                
## [55] fs_1.6.6                           lubridate_1.9.4                   
## [57] dplyr_1.1.4                        purrr_1.1.0                       
## [59] readr_2.1.5                        tidyr_1.3.1                       
## [61] tibble_3.3.0                       tidyverse_2.0.0                   
## [63] scooter_0.0.0.9004                 irlba_2.3.5.1                     
## [65] Matrix_1.7-0                       SeuratObject_5.2.0                
## [67] Seurat_4.4.0                       scater_1.34.1                     
## [69] scuttle_1.16.0                     SingleCellExperiment_1.28.1       
## [71] SummarizedExperiment_1.36.0        Biobase_2.66.0                    
## [73] GenomicRanges_1.58.0               GenomeInfoDb_1.42.3               
## [75] IRanges_2.40.1                     S4Vectors_0.44.0                  
## [77] BiocGenerics_0.52.0                MatrixGenerics_1.18.1             
## [79] matrixStats_1.5.0                  flexmix_2.3-20                    
## [81] lattice_0.22-6                     miQC_1.14.0                       
## [83] ggrepel_0.9.6                      ggpmisc_0.6.2                     
## [85] ggpp_0.5.9                         stringr_1.5.2                     
## [87] GGally_2.4.0                       ggplot2_4.0.0                     
## [89] forcats_1.0.1                      devtools_2.4.5                    
## [91] usethis_2.2.3                      cowplot_1.2.0                     
## [93] future_1.67.0                     
## 
## loaded via a namespace (and not attached):
##   [1] igraph_2.1.4                Formula_1.2-5              
##   [3] ica_1.0-3                   plotly_4.10.4              
##   [5] zlibbioc_1.52.0             tidyselect_1.2.1           
##   [7] bit_4.0.5                   rjson_0.2.21               
##   [9] blob_1.2.4                  urlchecker_1.0.1           
##  [11] S4Arrays_1.6.0              parallel_4.4.0             
##  [13] dichromat_2.0-0.1           seqLogo_1.72.0             
##  [15] png_0.1-8                   cli_3.6.5                  
##  [17] ggplotify_0.1.2             ProtGenerics_1.38.0        
##  [19] goftest_1.2-3               bluster_1.16.0             
##  [21] BiocNeighbors_2.0.1         uwot_0.2.2                 
##  [23] curl_7.0.0                  mime_0.12                  
##  [25] evaluate_0.24.0             tidytree_0.4.6             
##  [27] leiden_0.4.3.1              stringi_1.8.4              
##  [29] backports_1.5.0             XML_3.99-0.16.1            
##  [31] httpuv_1.6.15               magrittr_2.0.4             
##  [33] rappdirs_0.3.3              splines_4.4.0              
##  [35] jpeg_0.1-10                 sctransform_0.4.1          
##  [37] ggbeeswarm_0.7.2            sessioninfo_1.2.2          
##  [39] DBI_1.2.3                   HDF5Array_1.34.0           
##  [41] jquerylib_0.1.4             withr_3.0.2                
##  [43] xgboost_1.7.7.1             lmtest_0.9-40              
##  [45] BiocManager_1.30.23         biomaRt_2.62.1             
##  [47] SparseArray_1.6.2           annotate_1.84.0            
##  [49] VariantAnnotation_1.52.0    zoo_1.8-12                 
##  [51] UCSC.utils_1.2.0            TFMPvalue_0.0.9            
##  [53] timechange_0.3.0            caTools_1.18.3             
##  [55] ggtree_3.14.0               rhdf5_2.50.2               
##  [57] pwalign_1.2.0               quantreg_6.1               
##  [59] poweRlaw_0.80.0             alabaster.schemas_1.6.0    
##  [61] gridGraphics_0.5-1          ellipsis_0.3.2             
##  [63] lazyeval_0.2.2              survival_3.5-8             
##  [65] scattermore_1.2             BiocVersion_3.20.0         
##  [67] crayon_1.5.2                RcppAnnoy_0.0.22           
##  [69] progressr_0.14.0            later_1.3.2                
##  [71] base64enc_0.1-3             ggridges_0.5.6             
##  [73] codetools_0.2-20            profvis_0.3.8              
##  [75] KEGGREST_1.46.0             Rtsne_0.17                 
##  [77] limma_3.62.2                Rsamtools_2.22.0           
##  [79] filelock_1.0.3              foreign_0.8-86             
##  [81] pkgconfig_2.0.3             xml2_1.4.0                 
##  [83] GenomicAlignments_1.42.0    aplot_0.2.2                
##  [85] spatstat.sparse_3.0-3       alabaster.base_1.6.1       
##  [87] ape_5.8                     viridisLite_0.4.2          
##  [89] xtable_1.8-4                interp_1.1-6               
##  [91] plyr_1.8.9                  httr_1.4.7                 
##  [93] tools_4.4.0                 globals_0.18.0             
##  [95] pkgbuild_1.4.4              checkmate_2.3.1            
##  [97] htmlTable_2.4.2             beeswarm_0.4.0             
##  [99] nlme_3.1-164                ExperimentHub_2.14.0       
## [101] MatrixModels_0.5-4          lme4_1.1-35.3              
## [103] digest_0.6.35               farver_2.1.2               
## [105] tzdb_0.4.0                  yulab.utils_0.2.1          
## [107] viridis_0.6.5               rpart_4.1.23               
## [109] DirichletMultinomial_1.48.0 cachem_1.1.0               
## [111] polyclip_1.10-6             Hmisc_5.1-3                
## [113] generics_0.1.4              parallelly_1.45.1          
## [115] pkgload_1.3.4               statmod_1.5.0              
## [117] ScaledMatrix_1.14.0         minqa_1.2.7                
## [119] pbapply_1.7-2               httr2_1.2.1                
## [121] spam_2.10-0                 gson_0.1.0                 
## [123] dqrng_0.4.1                 gtools_3.9.5               
## [125] alabaster.se_1.6.0          shiny_1.11.1               
## [127] GenomeInfoDbData_1.2.13     rhdf5filters_1.18.1        
## [129] RCurl_1.98-1.14             memoise_2.0.1              
## [131] rmarkdown_2.27              scales_1.4.0               
## [133] gypsum_1.2.0                RANN_2.6.1                 
## [135] rstudioapi_0.16.0           spatstat.data_3.0-4        
## [137] cluster_2.1.6               spatstat.utils_3.0-4       
## [139] hms_1.1.3                   fitdistrplus_1.1-11        
## [141] colorspace_2.1-0            rlang_1.1.6                
## [143] DelayedMatrixStats_1.28.1   sparseMatrixStats_1.18.0   
## [145] dotCall64_1.1-1             ggtangle_0.0.7             
## [147] xfun_0.53                   alabaster.matrix_1.6.1     
## [149] CNEr_1.42.0                 remotes_2.5.0              
## [151] modeltools_0.2-23           abind_1.4-5                
## [153] GOSemSim_2.32.0             treeio_1.30.0              
## [155] ggsci_3.1.0                 Rhdf5lib_1.28.0            
## [157] bitops_1.0-7                promises_1.3.3             
## [159] RSQLite_2.3.7               qvalue_2.38.0              
## [161] fgsea_1.32.4                DelayedArray_0.32.0        
## [163] GO.db_3.20.0                compiler_4.4.0             
## [165] prettyunits_1.2.0           alabaster.ranges_1.6.0     
## [167] boot_1.3-30                 beachmat_2.22.0            
## [169] SparseM_1.84-2              polynom_1.4-1              
## [171] listenv_0.9.1               edgeR_4.4.2                
## [173] BiocSingular_1.22.0         tensor_1.5                 
## [175] progress_1.2.3              MASS_7.3-60.2              
## [177] BiocParallel_1.40.2         spatstat.random_3.2-3      
## [179] R6_2.6.1                    fastmap_1.2.0              
## [181] fastmatch_1.1-4             vipor_0.4.7                
## [183] ROCR_1.0-11                 ggstats_0.11.0             
## [185] rsvd_1.0.5                  nnet_7.3-19                
## [187] gtable_0.3.6                KernSmooth_2.23-22         
## [189] latticeExtra_0.6-30         miniUI_0.1.1.1             
## [191] deldir_2.0-4                htmltools_0.5.8.1          
## [193] bit64_4.0.5                 spatstat.explore_3.2-7     
## [195] lifecycle_1.0.4             S7_0.2.0                   
## [197] nloptr_2.0.3                restfulr_0.0.15            
## [199] sass_0.4.9                  vctrs_0.6.5                
## [201] VGAM_1.1-11                 spatstat.geom_3.2-9        
## [203] DOSE_4.0.1                  scran_1.34.0               
## [205] ggfun_0.2.0                 sp_2.1-4                   
## [207] future.apply_1.11.2         pracma_2.4.4               
## [209] bslib_0.7.0                 pillar_1.11.1              
## [211] metapod_1.14.0              locfit_1.5-9.9             
## [213] jsonlite_1.8.8
\end{verbatim}

\end{document}
